### ISSUE TITLE
`pl.date_range` doesn't support `date` subclasses with 1.6+

### ISSUE THREAD

**Issue by max-muoto** [pola-rs/polars#18613]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

With Polars 1.6, support for date subclasses in `pl.date_range` is broken. Take this example, which worked in prior versions:

```python
class mydate(dt.date): ...

pl.date_range(mydate(2022, 1, 1), mydate(2022, 3, 1), "1mo", eager=True).alias("date")
```

Which results in:

```
TypeError: cannot create expression literal for value of type mydate: mydate(2022, 1, 1)

Hint: Pass `allow_object=True` to accept any value and create a literal of type Object.
```

The same goes when using popular extensions on the `datetime` library such as [pendelum](https://pendulum.eustace.io/):

```python
pl.date_range(pendulum.instance(dt.date(2022, 1, 1)), pendulum.instance(dt.date(2022, 3, 1)), "1mo", eager=True).alias("date")
```

### Issue description

I didn't see this as a breaking change in 1.6, but if this is intended, feel free to close out this issue.

### Expected behavior

The expected behavior would be for the date ranges to generate in the same manner if non-subclass were being used.

```
In [6]: pl.date_range(dt.date(2022, 1, 1), dt.date(2022, 3, 1), "1mo", eager=True).alias("date")
   ...:
Out[6]:
shape: (3,)
Series: 'date' [date]
[
	2022-01-01
	2022-02-01
	2022-03-01
]
```

### Installed versions

<details>

```
--------Version info---------
Polars:              1.6.0
Index type:          UInt32
Platform:            macOS-14.4.1-arm64-arm-64bit
Python:              3.12.3 | packaged by conda-forge | (main, Apr 15 2024, 18:35:20) [Clang 16.0.6 ]

----Optional dependencies----
adbc_driver_manager  0.11.0
altair               <not installed>
cloudpickle          <not installed>
connectorx           <not installed>
deltalake            <not installed>
fastexcel            0.11.0
fsspec               <not installed>
gevent               <not installed>
great_tables         <not installed>
matplotlib           <not installed>
nest_asyncio         1.6.0
numpy                1.26.1
openpyxl             3.1.2
pandas               2.2.0
pyarrow              14.0.2
pydantic             2.8.2
pyiceberg            <not installed>
sqlalchemy           <not installed>
torch                <not installed>
xlsx2csv             0.8.2
xlsxwriter           3.1.9
```

</details>

**Comment by cmdlineluser** [comment]
I believe it was unintentional and was fixed after https://github.com/pola-rs/polars/issues/18447 was opened.

Your examples both run on main.

**Comment by max-muoto** [comment]
> I believe it was unintentional and was fixed after #18447 was opened.
> 
> Your examples both run on main.

Great, in that case I will close this out.
